lib/pengine/complex.c

```diff
@@ -435,6 +435,60 @@ detect_promotable(pe_resource_t *rsc)
     return FALSE;
 }
 
+static void
+free_params_table(gpointer data)
+{
+    g_hash_table_destroy((GHashTable *) data);
+}
+
+/*!
+ * \brief Get a table of resource parameters
+ *
+ * \param[in] rsc       Resource to query
+ * \param[in] node      Node for evaluating rules (NULL for defaults)
+ * \param[in] data_set  Cluster working set
+ *
+ * \return Hash table containing resource parameter names and values
+ *         (or NULL if \p rsc or \p data_set is NULL)
+ * \note The returned table will be destroyed when the resource is freed, so
+ *       callers should not destroy it.
+ */
+GHashTable *
+pe_rsc_params(pe_resource_t *rsc, pe_node_t *node, pe_working_set_t *data_set)
+{
+    GHashTable *params = NULL;
+
+    /* A NULL node is used to request the resource's default parameters
+     * (not evaluated for node), but we always want something non-NULL
+     * as a hash table key.
+     */
+    const char *node_name = "";
+
+    // Sanity check
+    if ((rsc == NULL) || (data_set == NULL)) {
+        return NULL;
+    }
+    if ((node != NULL) && (node->details->uname != NULL)) {
+        node_name = node->details->uname;
+    }
+
+    // Find the parameter table for given node
+    if (rsc->params == NULL) {
+        rsc->params = g_hash_table_new_full(crm_strcase_hash, crm_strcase_equal,
+                                            free, free_params_table);
+    } else {
+        params = g_hash_table_lookup(rsc->params, node_name);
+    }
+
+    // If none exists yet, create one with parameters evaluated for node
+    if (params == NULL) {
+        params = crm_str_table_new();
+        get_rsc_attributes(params, rsc, node, data_set);
+        g_hash_table_insert(rsc->params, strdup(node_name), params);
+    }
+    return params;
+}
+
```

Comment:
Not saying I have a really good idea for names but params, rsc->params and rsc->parameters being 3 different things is confusing. Maybe params_cache, params_node_cache ...

Reply:
Agreed, but as of this PR, ->parameters is deprecated and no longer used internally, and it would go away at the next API compatibility break.

I can rename the local params variable, though. Something like params_on_node or evaluated_params.

Reply:
I guess it does make sense to rename rsc->params as well since it's not obvious it's a table of tables. Maybe rsc->parameter_cache.